File: src/traust_engine/data/secure-code-audit/opengrep-rules/go/path-traversal.go
package rules

import (
	"archive/tar"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

func serveFile(w http.ResponseWriter, r *http.Request) {
	// ruleid: traust-go-path-traversal-taint
	os.ReadFile("/data/" + r.URL.Query().Get("name"))

	// ruleid: traust-go-path-traversal-taint
	http.ServeFile(w, r, filepath.Join("/reports", r.FormValue("file")))

	// ok: traust-go-path-traversal-taint
	os.ReadFile(filepath.Join("/data", filepath.Base(r.URL.Query().Get("name"))))

	// ok: traust-go-path-traversal-taint
	os.ReadFile("/etc/config/static.yaml")
}

func extract(dst string, hdr *tar.Header) {
	// ruleid: traust-go-path-traversal-tar-slip
	target := filepath.Join(dst, hdr.Name)
	_ = target

	if strings.HasPrefix(filepath.Clean(filepath.Join(dst, hdr.Name)), dst) {
		// ok: traust-go-path-traversal-tar-slip
		safe := filepath.Join(dst, hdr.Name)
		_ = safe
	}
}

File: src/traust_engine/data/secure-code-audit/opengrep-rules/python/path-traversal.py
import os
import tarfile
import zipfile


def unpack(archive_path, dest):
    tar = tarfile.open(archive_path)
    # ruleid: traust-python-path-traversal-tar-extractall-nofilter
    tar.extractall(dest)

    member = tar.getmembers()[0]
    # ruleid: traust-python-path-traversal-tar-extractall-nofilter
    tar.extract(member, dest)

    # ok: traust-python-path-traversal-tar-extractall-nofilter
    tar.extractall(dest, filter="data")

    # ok: traust-python-path-traversal-tar-extractall-nofilter
    tar.extract(member, dest, filter="data")


def unpack_zip(archive_path, dest):
    # zipfile sanitizes member paths on extract — zip-slip is tarfile-only;
    # flagging these re-seeds a refuted FP class (Precision Gate R9).
    zf = zipfile.ZipFile(archive_path)
    # ok: traust-python-path-traversal-tar-extractall-nofilter
    zf.extractall(dest)

    with zipfile.ZipFile(archive_path) as zf2:
        # ok: traust-python-path-traversal-tar-extractall-nofilter
        zf2.extractall(dest)


def contained(base, name):
    p = os.path.join(base, name)
    # ruleid: traust-python-path-traversal-containment-startswith
    if os.path.realpath(p).startswith(base):
        return p

    # ruleid: traust-python-path-traversal-containment-startswith
    if os.path.abspath(p).startswith(base):
        return p

    # ok: traust-python-path-traversal-containment-startswith
    if os.path.realpath(p).startswith(base + os.sep):
        return p

    # ok: traust-python-path-traversal-containment-startswith
    if os.path.abspath(p).startswith(os.path.join(base, "")):
        return p
    return None
